### PR TITLE
feat: Realtime inspector

### PR DESCRIPTION
**Pull request by ivasilov**

### CODE REVIEW

**Comment by saltcod** [comment]
I think the icons are still in transit? But just fyi, there are light mode versions of these 
screenshot-2023-10-31-at-22 23 20

**Comment by saltcod** [comment]
I think just channel instead of "Choose channel" ? 
screenshot-2023-10-31-at-22 24 16

**Comment by saltcod** [comment]
The main nav item links you to a blank 
https://studio-staging-git-feat-realtime-page-supabase.vercel.app/dashboard/project/_/realtime/ 

but the tool is at 
https://studio-staging-git-feat-realtime-page-supabase.vercel.app/dashboard/project/_/realtime/inspector

should the menu item link you right to /inspector?

File: studio/components/interfaces/Realtime/Inspector/EventsTable.tsx [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+  enabled: boolean
+  showSendEvent: () => void
+}) => {
+  const router = useRouter()
+  const { ref } = useParams()
+
+  return (
+    <div className="w-full max-w-md flex items-center flex-col">
+      {enabled ? <div>No Realtime events found</div> : null}
+      <div className="text-foreground-lighter">Realtime event logs will be shown here</div>
```
Comment (saltcod) [inline studio/components/interfaces/Realtime/Inspector/EventsTable.tsx:36]: When?
Maybe this should be the heading and the sub-heading should be a description or instruction. 

ie:

**Realtime event logs will be shown here** 
Choose a channel to get started showing logs.....

File: studio/components/interfaces/Realtime/Inspector/EventsTable.tsx [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+          </Button>
+        </div>
+        <div className="w-full px-5 py-4 items-center gap-4 inline-flex border-b bg-background">
+          <div className="grow flex-col flex">
+            <div className="text-foreground">Not sure what to do?</div>
+            <div className="text-foreground-lighter text-xs">Browse our documentation</div>
+          </div>
+          <Button type="default" iconRight={<IconExternalLink />}>
+            <a href="https://supabase.com/docs/guides/realtime" target="_blank">
+              <span>Docs</span>
```
Comment (saltcod) [inline studio/components/interfaces/Realtime/Inspector/EventsTable.tsx:82]: I _think_ we say "Documentation" for most links like this?

**Comment by saltcod** [comment]
These pills in figma are slightly thinner in the y dimension, and less padding in x dimension. 
screenshot-2023-10-31-at-22 29 21

**Comment by saltcod** [comment]
Maybe in this state, it could say something generic like "Event viewer" instead of "Select an event"?

screenshot-2023-10-31-at-22 31 28

**Comment by saltcod** [comment]
I thought I needed to select a channel in order to start listening? Not so? 

screenshot-2023-10-31-at-22 33 41

**Comment by saltcod** [comment]
This struck me as an input, not a dropdown. If there's only 2 options, maybe put in a radio group? (just an opinion though) 
screenshot-2023-10-31-at-22 38 26

File: studio/pages/project/[ref]/realtime/index.tsx [hunk 3]
```diff
@@ -0,0 +1,10 @@
+import RealtimeLayout from 'components/layouts/RealtimeLayout/RealtimeLayout'
```
Comment (joshenlim) [inline studio/pages/project/[ref]/realtime/index.tsx:1]: Probably can remove this file

**Review by alaister — COMMENTED** [review]
It would be nice to add some error state for this case:
Screenshot 2023-11-06 at 16 23 52
I didn't realise it wasn't working until I checked the network console!

File: studio/components/interfaces/Realtime/Inspector/useRealtimeMessages.ts [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+    (message: string, payload: any) => {
+      if (channel) {
+        channel.send({
+          type: 'broadcast',
+          event: message,
+          payload,
+        })
+        toast.success('Successfully broadcasted message')
+      } else {
+        toast.error('Failed to broadcast message: channel as not been set')
```
Comment (alaister) [inline studio/components/interfaces/Realtime/Inspector/useRealtimeMessages.ts:208]:
```suggestion
        toast.error('Failed to broadcast message: channel has not been set')
```
I'm also seeing this error when trying to send a test message with a channel set:
Screenshot 2023-11-06 at 16 27 43

File: studio/components/interfaces/Realtime/Inspector/Header.tsx [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+        <ChooseChannelPopover config={config} onChangeConfig={onChangeConfig} />
+        <Button
+          size="tiny"
+          type={config.enabled ? 'warning' : 'primary'}
+          className="rounded-l-none border-l-0"
+          disabled={config.channelName.length === 0}
+          icon={config.enabled ? <StopCircle size="16" /> : <PlayCircle size="16" />}
+          onClick={() => onChangeConfig({ ...config, enabled: !config.enabled })}
+        >
+          <span>{config.enabled ? `Stop listening` : `Start listening`}</span>
```
Comment (joshenlim) [inline studio/components/interfaces/Realtime/Inspector/Header.tsx:28]:
```suggestion
          {config.enabled ? `Stop listening` : `Start listening`}
```

File: studio/components/interfaces/Realtime/Inspector/MessagesFormatters.tsx [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+      </span>
+      <CopyButton
+        bounceIconOnCopy
+        text={value}
+        className="group-hover:opacity-100 opacity-0 my-auto transition col-span-2  h-4 w-4 px-0 py-0"
+        type="text"
+        title="Copy to clipboard"
+      >
+        {''}
+      </CopyButton>
```
Comment (joshenlim) [inline studio/components/interfaces/Realtime/Inspector/MessagesFormatters.tsx:35]:
```suggestion
      />
```

File: studio/components/interfaces/Realtime/Inspector/MessagesTable.tsx [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+}: {
+  enabled: boolean
+  showSendMessage: () => void
+}) => {
+  const router = useRouter()
+  const { ref } = useParams()
+
+  return (
+    <div className="w-full max-w-md flex items-center flex-col">
+      {enabled ? <div>No Realtime messages found</div> : null}
```
Comment (joshenlim) [inline studio/components/interfaces/Realtime/Inspector/MessagesTable.tsx:36]:
```suggestion
      {enabled ? <p>No Realtime messages found</p> : null}
```

File: studio/components/interfaces/Realtime/Inspector/MessagesTable.tsx [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+      <div className="text-foreground-lighter">Realtime message logs will be shown here</div>
```
Comment (joshenlim) [inline studio/components/interfaces/Realtime/Inspector/MessagesTable.tsx:37]:
```suggestion
      <p className="text-foreground-lighter">Realtime message logs will be shown here</p>
```

File: studio/components/interfaces/Realtime/Inspector/MessagesTable.tsx (hunk 7 above)
Comment (joshenlim) [reply studio/components/interfaces/Realtime/Inspector/MessagesTable.tsx:36]: if its just a text lets stick to p or span, avoid divs just to be semantically correct

File: studio/components/interfaces/Realtime/Inspector/RealtimeFilterPopover/AllSchemasSelector.tsx [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+  PopoverContent_Shadcn_,
+  PopoverTrigger_Shadcn_,
+  Popover_Shadcn_,
+  ScrollArea,
+} from 'ui'
+
+import { useProjectContext } from 'components/layouts/ProjectLayout/ProjectContext'
+import { useSchemasQuery } from 'data/database/schemas-query'
+
+interface SchemaSelectorProps {
```
Comment (joshenlim) [inline studio/components/interfaces/Realtime/Inspector/RealtimeFilterPopover/AllSchemasSelector.tsx:26]: Okay for now but can we extend the SchemaSelector instead? let's avoid duplicating code where possible

**Comment by joshenlim** [comment]
> It would be nice to add some error state for this case: Screenshot 2023-11-06 at 16 23 52 I didn't realise it wasn't working until I checked the network console!

was chatting to alaister briefly about this one - not sure how to repro, we can look into this another time

**Review by joshenlim — APPROVED** [review]

**Review by alaister — APPROVED** [review]

**Pull request merged**